adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java

```diff
@@ -286,6 +286,13 @@ void updateCachedItemWithResult(final String resource, final String clientId, fi
                 throw new AuthenticationException(ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_URL, e.getMessage(), e);
             }
         } else if (AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT.equalsIgnoreCase(result.getErrorCode())) {
+            // make sure to set authority to the one in preferred cache
+            try {
+                mAuthority = getAuthorityUrlWithPreferredCache();
+            }catch (MalformedURLException e){
+                //TODO : Understand what can be done here
+                Logger.d(TAG + methodName,  "Authority from preferred caches is invalid");
```

Comment:
This should not happen, but if MalformedURLException is thrown, we should throw AuthenticationException with invalid url error code.

Reply:
I agree this should never happen. But in this case we changing the authority url to preferred cache authority url and caller would have no knowledge about it. Not sure if throwing AuthenticationException is a right thing to do here as the url passed in by the caller is a valid one

Reply:
When storing the cache item, are we always using getAuthorityUrlWithPreferredCache() or using authority from request as well?

Reply:
Only if the InstanceDiscoveryMetadata of the authority is successful , we update it here : [AcquireTokenRequest.java#L258](https://github.com/AzureAD/azure-activedirectory-library-for-android/blob/962ce25365320e9eb58a90f7070f46520e2c5733/adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java#L258). However even if it fails, we could store the cache items with the request's authority url. But in this case `getAuthorityUrlWithPreferredCache()` would return us the request authority url itself , so I think we should be fine here.

However there is a case where we update the authority to cloudInstanceHostName here at [Oauth2.java#L287](https://github.com/AzureAD/azure-activedirectory-library-for-android/blob/962ce25365320e9eb58a90f7070f46520e2c5733/adal/src/main/java/com/microsoft/aad/adal/Oauth2.java#L287) and store the tokens in the cache with it. I am not sure if this case is handled as for the deletion to work `getAuthorityUrlWithPreferredCache()` should return the url of `cloudInstanceHostName. 

@iambmelt and @addev-ashish Correct em if I am wrong, but I think this could be potential issue here